### DISCUSSION TITLE
Creazione cliente. Errore Internal Server Error

### DISCUSSION THREAD

**Discussion by slear** [fattureincloud/api#188] [Q&A] [answered]
Buongiorno, sto provando a integrare le RestAPI per la gestione dei clienti nella mia applicazione. Purtroppo, ogni volta che provo a creare un cliente ottengo una risposta 500 Internal Server Error. Sto passando informazioni errate o incomplete? Eppure mi sembra che stia inserendo tutto quello che serve. Ecco un esempio del JSON inviato tramite una call eseguita in C# tramite RestSharp

        var url = "https://api-v2.fattureincloud.it";
        var api = $"/c/{mycompanyid}/entities/clients";

        var client = new RestClient(url);
        var request = new RestRequest(api, Method.Post);
        request.AddHeader("authorization", "Bearer " + token);
        request.AddJsonBody<ClienteInsertFIC>(ficInsert);
        RestResponse response = await client.ExecuteAsync(request);

ficInsert è una istanza della classe ClienteInsertFIC

        public class ClienteInsertFIC
        {
            public ClienteFICNew data { get; set; }
        }
        
        public class ClienteFICNew
        {
            public string code { get; set; }
            public string name { get; set; }
            public string type { get; set; }
            public string first_name { get; set; }
            public string last_name { get; set; }
            public string contact_person { get; set; }
            public string vat_number { get; set; }
            public string tax_code { get; set; }
            public string address_street { get; set; }
            public string address_postal_code { get; set; }
            public string address_city { get; set; }
            public string address_province { get; set; }
            public string address_extra { get; set; }
            public string country { get; set; }
            public string email { get; set; }
            public string certified_email { get; set; }
            public string phone { get; set; }
            public string fax { get; set; }
            public string notes { get; set; }
            public int default_payment_terms { get; set; }
            public string default_payment_terms_type { get; set; }
            public string bank_name { get; set; }
            public string bank_iban { get; set; }
            public string bank_swift_code { get; set; }
            public string shipping_address { get; set; }
            public bool e_invoice { get; set; }
            public string ei_code { get; set; }
            public Default_Vat default_vat { get; set; }
        }

I dati di esempio che posso mostrarvi sono i seguenti:

        ClienteFICNew ficNew = new ClienteFICNew
        {
            first_name = "Mario",
            last_name = "Rossi",
            name = "Mario Rossi",
            type = "person",
            address_city = "RIMINI",
            address_postal_code = "47924",
            address_street = "Via xxxxxxxxxx 2",
            email = "xxxx@xxxx.xx",
            phone = "+39xxxxxxxxxxxx",
            tax_code = "AAABBB99T01H294W",
            default_vat = new Default_Vat(),
            contact_person = "",
            vat_number = "",
            address_province = "",
            address_extra = "",
            country = "Italia",
            certified_email = "",
            fax = "",
            notes = "",
            default_payment_terms = 1,
            default_payment_terms_type = "standard",
        };
        ficInsert.data = ficNew;

Ma come detto mi ritorna sempre una Response con indicato un Internal Server Error 

        "Content": "{\"error\":{\"message\":\"An unexpected error occurred. Contact support for more information.\"}}",
        "StatusCode": 500,
        "IsSuccessStatusCode": false,
        "IsSuccessful": false,
        "StatusDescription": "Internal Server Error",

e naturalmente in cliente non viene aggiunto.

Grazie per l'aiuto che potrete fornirmi

**Comment by valmoz** [comment, marked as answer]
Buongiorno @slear 
le chiederei se possibile di inviarmi il JSON effettivo invece del codice C#.

In ogni caso, per la risoluzione di problemi suggeriamo sempre di scrivere al supporto clienti invece che alla community, in modo tale da poter indagare più a fondo sul suo problema.

P.S.: qualora le fosse possibile, suggerirei di utilizzare la [C# SDK](https://developers.fattureincloud.it/docs/sdks/csharp-sdk)

Grazie
Mauro

**Comment by slear** [comment]
Grazie Mauro, ora provvedo a fare una richiesta al supporto clienti.
In ogni caso, il json inviato nella request è il seguente (ovviamente con nomi di fantasia, ma nell'invio il codice fiscale corrisponde al nominativo effettivo)
Forse dovrei provare a mettere delle stringhe vuote al posto di quei null anche se non sono campi obbligatori

          {
            "data": {
              "code": "PZ0001",
              "name": "Mario Rossi",
              "type": "person",
              "first_name": "Mario",
              "last_name": "Rossi",
              "contact_person": "",
              "vat_number": "",
              "tax_code": "AAAVVV99T01H294W",
              "address_street": "Via XXXXXXXXX 2",
              "address_postal_code": "47924",
              "address_city": "RIMINI",
              "address_province": "",
              "address_extra": "",
              "country": "Italia",
              "email": "XXXXX@XXXX.it",
              "certified_email": "",
              "phone": "XXXXXXXXXX",
              "fax": "",
              "notes": "",
              "default_payment_terms": 1,
              "default_payment_terms_type": "standard",
              "bank_name": null,
              "bank_iban": null,
              "bank_swift_code": null,
              "shipping_address": null,
              "e_invoice": false,
              "ei_code": null,
              "default_vat": {
                "id": 0,
                "value": 0,
                "description": null,
                "is_disabled": false
              }
            }
          }

**Reply by valmoz** [reply]
Ciao @slear 
suggerirei di provare a rimuovere totalmente quei campi null. Solitamente le librerie JSON hanno l'opzione per ignorare tali campi in fase di marshalling.

Mauro

**Comment by slear** [comment]
Era proprio quello Mauro. Non ho modificato le opzioni per la serializazzione del Json, ma ho semplicemente inizializzato con stringa vuota i campi mancanti e adesso non mi da più errore. Anche se non mi restituisce l'id generato per il nuovo cliente. Ma questa è un'altra storia, Grazie, spero che questa chiacchierata sia utile alla community.
